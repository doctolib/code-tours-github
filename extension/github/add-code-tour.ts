import 'showdown-prettify'
import * as showdown from 'showdown'
import { filterXSS } from 'xss'
import { forwardRequest } from './forward-request'
import { EnhancedCodeTourStep } from '../types/code-tour'

const converter = new showdown.Converter({ extensions: ['prettify'] })
converter.setFlavor('github')

async function getStep(title: string, step: number): Promise<EnhancedCodeTourStep> {
  const response = await forwardRequest({ action: 'GET_STEP', codeTourStep: step, codeTourTitle: title })
  if (response.action !== 'STEP') throw new Error('should return a step')
  return response.step
}

function buttonTo(text: string, url?: string): Element {
  const classes = url ? 'btn btn-primary' : 'btn btn-primary disabled'
  const button = document.createElement('a')
  button.setAttribute('class', classes)
  button.setAttribute('style', 'margin-top: 10px; margin-right: 5px;')
  button.setAttribute('href', url || '')
  button.text = text

  return button
}

function getParent(currentStep: EnhancedCodeTourStep) {
  if ('file' in currentStep) {
    const currentLine = currentStep.line
    return document.querySelector(`#LC${currentLine}.blob-code`)
  }
  return document.querySelector('div.repository-content')
}

function formatAndSanitizeDescription(rawText: string): string {
  return filterXSS(converter.makeHtml(rawText))
}

function buildTitleRow(currentStep: EnhancedCodeTourStep, stepNumber: number) {
  const titleRow = document.createElement('p')

  const img = document.createElement('img')
  img.setAttribute('src', chrome.extension.getURL('code-tour.png'))
  img.setAttribute('style', 'height: 2em; margin-right: 1em;')

  const codeTour = document.createElement('b')
  codeTour.textContent = 'CodeTour'

  const hr = document.createElement('hr')

  const currentTourInfo = document.createElement('span')
  currentTourInfo.textContent = ` Step ${stepNumber + 1} of ${currentStep.tour.steps.length} (${
    currentStep.tour.title
  })`

  titleRow.append(img, codeTour, currentTourInfo, hr)
  return titleRow
}

export async function addCodeTour(): Promise<void> {
  const sheet = document.createElement('style')
  sheet.innerHTML = `
  pre {
    border: 1px black solid;
    padding: 1em;
  }
  `
  document.body.prepend(sheet)

  const searchParams = new URLSearchParams(window.location.search)
  const name = searchParams.get('code-tour')
  const step = parseInt(searchParams.get('step') ?? '', 10) || 0

  if (!name) return

  const currentStep = await getStep(name, step)
  const currentDescription = formatAndSanitizeDescription(currentStep.description)

  const previousButton = buttonTo('Previous', currentStep.previousUrl)
  const nextButton = buttonTo('Next', currentStep.nextUrl)

  const section = document.createElement('div')
  section.setAttribute('class', 'dl-doctolib-code-tour-comment')

  const tourInfo = buildTitleRow(currentStep, step)

  const span = document.createElement('span')
  span.innerHTML = currentDescription

  const br = document.createElement('br')

  section.append(tourInfo, span, br, previousButton, nextButton)

  section.setAttribute(
    'style',
    `
    padding: 14px;
    margin: 14px;
    border: 1px lightgrey solid;
    background-color: white;
    border-radius: 1em;
    font-family: sans-serif;
    `,
  )

  const parent = getParent(currentStep)
  if (!parent) return
  if ('file' in currentStep) {
    parent.append(section)
    parent.classList.add('highlighted')
    parent.scrollIntoView({ behavior: 'auto', block: 'center' })
  } else {
    parent.prepend(section)
  }
}
