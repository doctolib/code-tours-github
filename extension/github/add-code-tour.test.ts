import { buildSection, buildTitleRow, getParent } from './add-code-tour'

global.chrome = {
  // @ts-expect-error mocking the chrome API in tests
  extension: {
    getURL: (element) => `extension-url/${element}`,
  },
}

describe(buildTitleRow, () => {
  it('renders properly the code tour header', () => {
    expect(
      buildTitleRow(
        {
          description: 'code tour description',
          line: 42,
          file: '/package.json',
          nextUrl: 'the/next/url?with=params',
          previousUrl: 'the/next/url?with=params',
          tour: {
            title: 'The code tour title',
            step: 3,
            steps: [],
            ref: 'alongcommitid',
            repository: 'doctolib/code-tours-github',
          },
        },
        1,
      ),
    ).toMatchSnapshot()
  })
})

describe(buildSection, () => {
  it('renders properly the whole section', () => {
    expect(
      buildSection(
        {
          description: 'code tour description',
          line: 42,
          file: '/package.json',
          nextUrl: 'the/next/url?with=params',
          previousUrl: 'the/next/url?with=params',
          tour: {
            title: 'The code tour title',
            step: 3,
            steps: [],
            ref: 'alongcommitid',
            repository: 'doctolib/code-tours-github',
          },
        },
        1,
      ),
    ).toMatchSnapshot()
  })

  it('renders properly the section when the next and previous URLs are not defined', () => {
    expect(
      buildSection(
        {
          description: 'code tour description',
          line: 42,
          file: '/package.json',
          tour: {
            title: 'The code tour title',
            step: 3,
            steps: [],
            ref: 'alongcommitid',
            repository: 'doctolib/code-tours-github',
          },
        },
        1,
      ),
    ).toMatchSnapshot()
  })
})

describe(getParent, () => {
  it('returns the line of the file for a file step', () => {
    document.body.innerHTML = `
     <div id="LC42" class="blob-code" />
    `
    expect(
      getParent({
        description: 'code tour description',
        line: 42,
        file: '/package.json',
        tour: {
          title: 'The code tour title',
          step: 3,
          steps: [],
          ref: 'alongcommitid',
          repository: 'doctolib/code-tours-github',
        },
      }),
    ).toMatchSnapshot()
  })

  it('returns the top of the file for step with no file associated', () => {
    document.body.innerHTML = `
     <div class="repository-content"/>
    `
    expect(
      getParent({
        description: 'code tour description',
        directory: 'the directory',
        tour: {
          title: 'The code tour title',
          step: 3,
          steps: [],
          ref: 'alongcommitid',
          repository: 'doctolib/code-tours-github',
        },
      }),
    ).toMatchSnapshot()
  })
})
