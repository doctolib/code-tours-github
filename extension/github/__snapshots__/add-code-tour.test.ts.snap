// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildSection renders properly the section when the next and previous URLs are not defined 1`] = `
<div
  class="dl-doctolib-code-tour-comment"
  style="
    padding: 14px;
    margin: 14px;
    border: 1px lightgrey solid;
    background-color: white;
    border-radius: 1em;
    font-family: sans-serif;
    "
>
  <p
    style="display: flex; align-items: center;"
  >
    <img
      src="extension-url/code-tour.png"
      style="height: 2em; margin-right: 1em;"
    />
    <b>
      CodeTour
    </b>
    <span>
       Step 2 of 0 (The code tour title)
    </span>
    <hr />
  </p>
  <span>
    <p>
      code tour description
    </p>
  </span>
  <br />
  <a
    class="btn btn-primary disabled"
    href=""
    style="margin-top: 10px; margin-right: 5px;"
  >
    Previous
  </a>
  <a
    class="btn btn-primary disabled"
    href=""
    style="margin-top: 10px; margin-right: 5px;"
  >
    Next
  </a>
</div>
`;

exports[`buildSection renders properly the whole section 1`] = `
<div
  class="dl-doctolib-code-tour-comment"
  style="
    padding: 14px;
    margin: 14px;
    border: 1px lightgrey solid;
    background-color: white;
    border-radius: 1em;
    font-family: sans-serif;
    "
>
  <p
    style="display: flex; align-items: center;"
  >
    <img
      src="extension-url/code-tour.png"
      style="height: 2em; margin-right: 1em;"
    />
    <b>
      CodeTour
    </b>
    <span>
       Step 2 of 0 (The code tour title)
    </span>
    <hr />
  </p>
  <span>
    <p>
      code tour description
    </p>
  </span>
  <br />
  <a
    class="btn btn-primary"
    href="the/next/url?with=params"
    style="margin-top: 10px; margin-right: 5px;"
  >
    Previous
  </a>
  <a
    class="btn btn-primary"
    href="the/next/url?with=params"
    style="margin-top: 10px; margin-right: 5px;"
  >
    Next
  </a>
</div>
`;

exports[`buildTitleRow renders properly the code tour header 1`] = `
<p
  style="display: flex; align-items: center;"
>
  <img
    src="extension-url/code-tour.png"
    style="height: 2em; margin-right: 1em;"
  />
  <b>
    CodeTour
  </b>
  <span>
     Step 2 of 0 (The code tour title)
  </span>
  <hr />
</p>
`;

exports[`getParent returns the line of the file for a file step 1`] = `
<div
  class="blob-code"
  id="LC42"
>
  
    
</div>
`;

exports[`getParent returns the top of the file for step with no file associated 1`] = `
<div
  class="repository-content"
>
  
    
</div>
`;
